src/hashes.rs

```diff
@@ -1,473 +1,313 @@
+                fn input(&mut self, data: &[u8]) {
+                    self.0.input(data)
+                }
+                #[inline]
+                fn result(self) -> Multihash {
+                    wrap(Self::CODE, self.0.result().as_slice())
+                }
+                #[inline]
+                fn result_reset(&mut self) -> Multihash {
+                    wrap(Self::CODE, self.0.result_reset().as_slice())
+                }
+                #[inline]
+                fn reset(&mut self) {
+                    self.0.reset()
+                }
+            }
+            impl ::std::io::Write for $name {
+                #[inline]
+                fn write(&mut self, buf: &[u8]) -> ::std::io::Result<usize> {
+                    <$name as MultihashDigest>::input(self, buf);
+                    Ok(buf.len())
+                }
+                #[inline]
+                fn flush(&mut self) -> ::std::io::Result<()> {
+                    Ok(())
+                }
+            }
+        )*
+    };
+    ($(
+        #[$doc:meta]
+        @blake $type:ty | $params:ty as $name:ident $len:expr;
+            @code_doc $code_doc:literal,
+    )*) => {
+        $(
+            #[$doc]
+            #[derive(Clone, Debug)]
+            pub struct $name($type);
+            impl $name {
+                #[doc = $code_doc]
+                pub const CODE: Code = Code::$name;
+                /// Hash some input and return the Multihash digest.
+                pub fn digest(data: &[u8]) -> Multihash {
+                    let digest = <$params>::new().hash_length($len).hash(data);
+                    wrap(Self::CODE, &digest.as_bytes())
+                }
+            }
+            impl Default for $name {
+                fn default() -> Self {
+                    $name(<$params>::new().hash_length($len).to_state())
+                }
+            }
+            impl MultihashDigest for $name {
+                #[inline]
+                fn code(&self) -> Code {
+                    Self::CODE
+                }
+                #[inline]
+                fn digest(&self, data: &[u8]) -> Multihash {
+                    Self::digest(data)
+                }
+                #[inline]
+                fn input(&mut self, data: &[u8]) {
+                    self.0.update(data);
+                }
+                #[inline]
+                fn result(self) -> Multihash {
+                    let digest = self.0.finalize();
+                    wrap(Self::CODE, digest.as_bytes())
+                }
+                #[inline]
+                fn result_reset(&mut self) -> Multihash {
+                    let digest = self.0.finalize();
+                    let hash = wrap(Self::CODE, digest.as_bytes());
+                    self.reset();
+                    hash
+                }
+                #[inline]
+                fn reset(&mut self) {
+                    self.0 = <$params>::new().hash_length($len).to_state();
+                }
```

Comment:
Ok, my latest commit adds `result_reset` and `reset`, which for the Blake digests, just replaces the inner `blake2#_simd` digest with a new one.